shakenfist/node.py

```diff
@@ -0,0 +1,112 @@
+from functools import partial
+import time
+
+from shakenfist import baseobject
+from shakenfist import etcd
+from shakenfist import logutil
+from shakenfist import util
+
+
+LOG, _ = logutil.setup(__name__)
+
+
+class Node(baseobject.DatabaseBackedObject):
+    object_type = 'node'
+    current_version = 2
+    state_targets = {
+        None: ('created', 'error', 'missing'),
+        'created': ('deleted', 'error', 'missing'),
+        'error': ('error', 'deleted'),
+        'missing': ('created', 'error')
+    }
+
+    # docs/development/state_machine.md has a description of these states.
+    STATE_CREATED = 'created'
+    STATE_ERROR = 'error'
+    STATE_DELETED = 'deleted'
+    STATE_MISSING = 'missing'
+
+    def __init__(self, static_values):
+        # We treat a node name as a UUID here for historical reasons
+        super(Node, self).__init__(static_values['fqdn'],
+                                   static_values.get('version'))
+
+        self.__ip = static_values['ip']
+
+    @classmethod
+    def new(cls, name, ip):
+        n = Node.from_db(name)
+        if n:
+            return n
+
+        Node._db_create(name, {
+            'fqdn': name,
+            'ip': ip,
+            'version': cls.current_version
+        })
+        n = Node.from_db(name)
+        n.state = cls.STATE_CREATED
+        n.add_event('db record creation', None)
+        return n
+
+    @staticmethod
+    def from_db(fqdn):
+        if not fqdn:
+            return None
+
+        static_values = Node._db_get(fqdn)
+        if not static_values:
+            return None
+
+        return Node(static_values)
+
+    def external_view(self):
+        # If this is an external view, then mix back in attributes that users
+        # expect
+        return {
+            'uuid': self.uuid,
+            'fqdn': self.fqdn,
+            'ip': self.ip,
+            'state': self.state.value,
+            'lastseen': self.last_seen,
+            'release': self.installed_version
+        }
+
+    # Static values
+    @property
+    def ip(self):
+        return self.__ip
+
+    # Values routed to attributes, writes are via helper methods.
+    @property
+    def last_seen(self):
+        return self._db_get_attribute('observed')['at']
+
+    @property
+    def installed_version(self):
+        return self._db_get_attribute('observed')['release']
+
+    def observe(self):
```

Comment:
What do you think about making this a static method. It could call new() itself.

This would remove the need for the "upsert" explanations in other parts of the code.

Maybe also change the method name to past tense?  `observed`

Reply:
So, I tweaked this along those lines, but I've ended up calling it observe_this_node() because it will only work for the node the code is running on.